wurst/network/Network.wurst

```diff
@@ -0,0 +1,922 @@
+
+function precomputeKeys(string prefix, integer depth)
+	if depth == 0
+		for i = 0 to GC_KEYS.length() - 1
+			gcKeys[gcKeysCount] = prefix + GC_KEYS.charAt(i)
+			gcKeysCount++
+	else
+		for i = 0 to GC_KEYS.length() - 1
+			execute(() -> precomputeKeys(prefix + GC_KEYS.charAt(i), depth - 1))
+
+function precomputeKeys(integer length)
+	precomputeKeys("", length - 1)
+
+function precomputeLookup()
+	alphabetLookup = new Table()
+
+	for i = 0 to STRING_ALPHABET.length() - 1
+		let c = STRING_ALPHABET.charAt(i)
+
+		if (c.toUpperCase() == c and c.toLowerCase() != c) or c == "/"
+			alphabetLookup.saveInt(("00" + c).getHash(), i)
+		else
+			alphabetLookup.saveInt(c.getHash(), i)
+
+function testCollisions()
+	for i = 0 to STRING_ALPHABET.length() - 1
+		if getIndexFromChar(getCharFromIndex(i)) != i
+			print("CollisionTest failed for " + i.toString() + " " + getCharFromIndex(i))
+			print("getIndexFromChar() == " + getIndexFromChar(getCharFromIndex(i)).toString())
+			print("getCharFromIndex(getIndexFromChar)) == " + getCharFromIndex(getIndexFromChar(getCharFromIndex(i))))
+
+function getKey(integer i) returns string
+	if i >= gcKeysCount
+		error("Network: trying to get invalid GC key")
+	return gcKeys[i]
+
+function getCharFromIndex(integer i) returns string
+	if i > STRING_ALPHABET.length()
+		error("Network: trying to get invalid character from index '" + i.toString() + "'")
+	return STRING_ALPHABET.charAt(i)
+
+function getIndexFromChar(string c) returns integer
+	integer hash
+	if (c.toUpperCase() == c and c.toLowerCase() != c) or c == "/"
+		hash = ("00" + c).getHash()
+	else
+		hash = c.getHash()
+
+	if SAFETY_CHECKS_ENABLED and not alphabetLookup.hasInt(hash)
+		error("Network: trying to get index from invalid character '" + c + "'")
+
+	return alphabetLookup.loadInt(hash)
+
+init
+	for i = 1 to MAX_CHARACTERS_PER_KEY - 1
+		execute(() -> precomputeKeys(i))
+	execute(() -> precomputeLookup())
+	execute(() -> testCollisions())
+
+// simple box class to allow modifying a single value inside a closure
+class Box<T>
+	T value
+
+	construct(T value)
+		this.value = value
+
+class BoxBoolean
+	boolean value
+
+	construct(boolean value)
+		this.value = value
+
+class BoxString
+	string value
+
+	construct(string value)
+		this.value = value
+
+/**
+	Interface to be implemented by classes that wish to be directly
+	read/written to Network
+*/
+public interface Serializable
+	function serialize(DataStream stream)
+	function deserialize(DataStream stream)
+
+enum StreamType
+	INTEGER
+	REAL
+	BOOLEAN
+
+enum StreamMode
+	WRITING
+	READING
+	LOCKED
+
+/**
+	This class is used as an intermediate buffer to store data before it is synchronized.
+	That way, we can send data through gamecache using multiple bursts, while keeping the
+	gamecache key lengths short.
+**/
+public class DataStream
+	// max characters to store per compacting round
+	private static constant integer CHARS_PER_COMPACT = 512
+
+	private Table buffer = new Table()
```

Comment:
can these Tables be replaced by HashMaps?

Reply:
The choice is between 1 Table and 4 HashMaps. HashMaps: Reals cannot be cast to ints safely without losing precision. Strings require an additional store in another table. I think a single Table here is justified.

Would totally use HashMaps if not for the above issue.

Reply:
in this case it's ok